addons/mail/models/mail_message.py

```diff
@@ -417,13 +418,14 @@ def message_format(self):
         message_tree = dict((m.id, m) for m in self.sudo())
         self._message_read_dict_postprocess(message_values, message_tree)
 
-        # add subtype data (is_note flag, subtype_description). Do it as sudo
+        # add subtype data (is_note flag, is_discussion flag , subtype_description). Do it as sudo
         # because portal / public may have to look for internal subtypes
         subtype_ids = [msg['subtype_id'][0] for msg in message_values if msg['subtype_id']]
-        subtypes = self.env['mail.message.subtype'].sudo().browse(subtype_ids).read(['internal', 'description'])
+        subtypes = self.env['mail.message.subtype'].sudo().browse(subtype_ids).read(['internal', 'description','id'])
         subtypes_dict = dict((subtype['id'], subtype) for subtype in subtypes)
         for message in message_values:
             message['is_note'] = message['subtype_id'] and subtypes_dict[message['subtype_id'][0]]['internal']
+            message['is_discussion'] = message['subtype_id'] and subtypes_dict[message['subtype_id'][0]]['id'] == self.env['ir.model.data'].xmlid_to_res_id('mail.mt_comment')
```

Comment:
Two things for this line :

 - You can use self.env.ref() that is a crapper to xmlid_to_res_id. This one is a bit outdated and comes from old versions of Odoo. env.ref() is now commonly used when fetching data from an xml ID.
 - You can extract the call to xmlid_to_res_id (or env.ref) above the loop. There is no real performance impact as for this specific method the result is put into a cache and therefore no extra query is performed but it is a good practice.

Reply:
Thanks I will check that.